text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,316 @@
+
+    POST /participation/v1/joinbysnapshot
+
+ or the following CLI command:
+
+    $ osnadmin channel joinbysnapshot -s <snapshot_dir>
+
+There is no need to specify {channelID} when using the above API, as it will be extracted from the snapshot content.
+
+To check the status of the "join by snapshot" operation use the following REST API:
+
+    GET /participation/v1/channels/{channelID}
+
+ or the following CLI command:
+
+    $ osnadmin channel list -c <channelID>
+
+## Guide: Prune block ledger of an existing channel
+[guide-prune-block-ledger-of-an-existing-channel]: #guide-prune-block-ledger-of-an-existing-channel
+
+To prune an existing channel ledger do the following:
+ - create a new snapshot at the desired block number,
+ - remove the OSN from the channel,
+ - join the channel from the snapshot.
+
+This approach has some drawbacks, see [Drawbacks section](#drawbacks).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Changes to the existing API
+
+This RFC adds a new endpoint to the existing **channel participation API** to allow joining a channel by snapshot. Additionally this RFC defines new **ledger management API** and introduces a new endpoint of this API to allow snapshot creation. In future we plan to add at least one more endpoint to this API which will enable ledger pruning "on the fly" without interruption in OSN service to resolve the known [drawbacks](#drawbacks).
+
+Technically we achieve the above by following these steps:
+1. Introduce new REST API endpoint **/participation/v1/joinbysnapshot** while keeping all existing functionality of the channel participation API, that is keeping "join by join-block", "remove channel", and "list channel(s)" API endpoints.
+2. Add new REST API handler for the ledger management API with prefix **/ledger/v1/**. Define a single endpoint of this API **/ledger/v1/{channelID}/snapshot** to allow snapshot creation.
+3. Add "Ledger Management API Configuration" section in the OSN YAML config file.
+4. Add support of two new snapshot operations to the CLI tool *osnadmin*, specifically the commands *osnadmin channel joinbysnapshot* and *osnadmin ledger snapshot*.
+
+## Snapshot content
+
+We propose the OSN snapshot to follow the same logical structure as peer snapshot, that is to comprise a directory with a number of files in it, one of the files being a metadata JSON file. But here the similarity ends, as the content of the files will be drastically different.
+Here is the content of the OSN snapshot:
+1. Snapshot version - to allow for the future extensions. Current version is 1.
+2. Channel ID.
+3. lastBlockNum - the number of the last pruned block.
+4. last config block before lastBlockNum - this is required to verify the newly pulled blocks. The sequence number of the last config block must be lower or equal to lastBlockNum.
```

Comment:
> last config block before lastBlockNum

I don't understand. If there is a config block before the "lastBlockNum" then that config block is essentially the last block, no? 

Also, "the number of the last pruned block" - Shouldn't we care about the number of the lowest block that is not pruned?

Reply:
> I don't understand. If there is a config block before the "lastBlockNum" then that config block is essentially the last block, no?

No. It may happen that the current block has sequence number 1,000,000 and the last config block has the sequence number 100.
We want to be able to prune the ledger at 1 million sequence number to save the disk space.

> Also, "the number of the last pruned block" - Shouldn't we care about the number of the lowest block that is not pruned?

"the number of the lowest block that is not pruned" =  lastBlockNum + 1

it does not matter what number to store